examples/rcn.py

```diff
@@ -0,0 +1,452 @@
+
+# Adding rcn model edges to the pgmax factor graph.
+for idx in range(edges.shape[0]):
+    edge = edges[idx]
+
+    for e in edge:
+        i1, i2, r = e
+        fg.add_factor(
+            [(idx, i1), (idx, i2)],
+            valid_configs_list[r],
+        )
+
+end = time.time()
+print(f"Creating factors took {end-start:.3f} seconds.")
+
+
+# %% [markdown]
+# # 5. Run inference
+
+# %% [markdown]
+# ## 5.1 Helper functions to initialize the evidence for a given image
+
+# %%
+def get_bu_msg(img: np.ndarray) -> np.ndarray:
+    """Computes the bottom-up messages given a test image.
+
+    Args:
+        img: The rgb image to compute bottom up messages on [H, W, 3].
+
+    Returns:
+        An array of shape [16, H, W] denoting the presence or absence of an oriented and directional
+            line-segments at a particular location. The elements of this array belong to the set {+1, -1}.
+    """
+
+    # Convolving the image with different gabor filters.
+    filtered = np.zeros((filters.shape[0],) + img.shape, dtype=np.float32)
+    for i in range(filters.shape[0]):
+        kern = filters[i, :, :]
+        filtered[i] = fftconvolve(img, kern, mode="same")
+
+    # Apllying non-max supression to all the filtered images.
+    localized = np.zeros_like(filtered)
+    cross_orient_max = filtered.max(0)
+    filtered[filtered < 0] = 0
+    for i, (layer, suppress_mask) in enumerate(zip(filtered, suppression_masks)):
+        competitor_maxs = maximum_filter(layer, footprint=suppress_mask, mode="nearest")
+        localized[i] = competitor_maxs <= layer
+    localized[cross_orient_max > filtered] = 0
+
+    # Threshold and binarize
+    localized *= (filtered / brightness_diff_threshold).clip(0, 1)
+    localized[localized < 1] = 0
+
+    # Apply cross-channel pooling.
+    pooled_channels = [-np.ones_like(sf) for sf in localized]
+    for i, pc in enumerate(pooled_channels):
+        for channel_offset in [0, -1, 1]:
+            ch = (i + channel_offset) % num_orients
+            pos_chan = localized[ch]
+            np.maximum(pc, pos_chan, pc)
+
+    # Remapping the elements to set {+1, -1}.
+    bu_msg = np.array(pooled_channels)
+    bu_msg[bu_msg == 0] = -1
+    return bu_msg
+
+
+# %% [markdown]
+# ## 5.1.1 Visualizing bu_msg for a sample image
+
+# %% [markdown]
+# bu_msg has shape (16, H, W) where each 1 <= f <= 16 denotes the present or absense of a oriented edge
+
+# %%
+r_test_img = test_set[4]
+r_bu_msg = get_bu_msg(r_test_img)
+img = np.ones((200, 200))
+fig, ax = plt.subplots(1, 2, figsize=(20, 10))
+ax[0].imshow(r_test_img, cmap="gray")
+ax[0].axis("off")
+ax[0].set_title("Input image", fontsize=20)
+for i in range(r_bu_msg.shape[0]):
+    img[r_bu_msg[i] > 0] = 0
+
+ax[1].imshow(img, cmap="gray")
+ax[1].axis("off")
+ax[1].set_title("Max filter response across 16 channels", fontsize=20)
+fig.tight_layout()
+
+
+# %% [markdown]
+# ## 5.2 Run maximum a-posteriori probability (map) inference on all test images
```

Comment:
I believe the word probability is redundant here

Reply:
My earlier comment was intended for potential typo in usage of "map product inference". 
But you can just use MAP inference here without having to expand.

Reply:
Changed to say MAP inference.